doc/playonlinux.1
.TH PLAYONLINUX 1 "February  6, 2009"

.SH NAME
PlayOnLinux \- front-end for Wine

.SH SYNOPSIS
.B playonlinux
.RI [ options ]

.SH DESCRIPTION
.B PlayOnLinux
is a front-end for Wine. It allows you to easily install and use numerous games and softwares designed to run with Microsoft®'s Windows®.
.br
Few games are compatible with GNU/Linux at the moment and it certainly is a factor preventing the migration to this system.
.P
.B PlayOnLinux
performs installations in
.I ~/.PlayOnLinux
and uses separate Wine prefixes to avoid conflicts between applications.

.SH OPTIONS
.TP
.I --version
 Show version number and exit.
.TP
.I "--run <prog>"
 Run directly the specified program.

.SH AUTHOR
playonlinux was written by Quentin Pâris <qparis@playonlinux.com>.
.PP
This manual page was written by Bertrand Marc <beberking@gmail.com>,
for the Debian project (but may be used by others).


CHANGELOG
-> 4.2.1
- When removing shortcuts or virtual drives from the Configure window,
  make sure desktop icons, menu entries, etc. are also removed.
- fix Python's VersionLower (infinite source of bugs)
- Update © in about box LP: #1160801
- Fix a huge problem in bug reporting I introduced in 4.2... I thought
  it tested it well, this is depressing. Without testers, without
  users testing beta versions, it seems there's no way to get bugfree
  releases.
- (experimental) enable display of beta scripts by default,
  differenciate them in lists using a reddish background
- read script lists in utf8 when looking up script to install (#2076)

-> 4.2
- Huge icons download speedup by reusing a single HTTP connection
- Make sure debug mode is disabled when using playonlinux --run; Debug may still be used in playonlinux-bash though
- Make sure entered prefix name contains no space or slash during manual installations
- POL_System_wget: make sure the pipelined wget exitcode is not masked by grep's one
- POL_System_wget: interpret wget exitcodes
- genere_icone(): pick the right icon instead of Wine generic one (reported by Simon/excalibr on IRC) (#1815)
- (experimental) WINEDEBUG prefix setting, modify the amount of   logging when debugging is enabled. I'm not sure of the right scope,
  should it be per-shortcut, like the old "enable debugging" flag, or even global?
- POL_SetupWindow_shortcut_creator: use executable name as default shortcut name
- sort shortcuts case insensitively in main window so that lowercase doesn't end up at the bottom
- bug report: insert basic questions to answer in the bug report
- add "Report a problem" in shortcut-related actions in the side panel
- do not ask for confirmation to close PlayOnLinux if there's no program started
- do not suggest bug reporting if a script has been canceled by the user because a prefix already existed
- virtual drive removal should now overpass bad directory rights (caused by WMP9 for example) (#1732)
- POL_System_HomeSpaceLeft fix for df multiline output (#1890)
- added POL_System_UserRootSpaceLeft, in case ~/.PlayOnLinux is a symlink or a mount
- use POL_System_UserRootSpaceLeft in POL_System_EnoughSpace
- if exists, look for mounted devices in /media/$USER/ directory (Ubuntu Quantal, #1893)
- manual_install: don't call POL_Wine_PrefixCreate on patch installation
- irc: only warn about closing window when connected
- Debug log all user inputs (POL_SetupWindow_textbox, etc.)
- mainwindow.py: try to handle EINTR while reading from pol_update_list pipe (#1687)
- DeletePrefix: if root is actually a symlink, just delete the symlink
- gui_server.py: try to handle EINTR while accepting socket connections (#1687)
- Fix typo (#1932)
- POL_SetupWindow_download: accept a directory or a filename as 4th parameter
- POL_Download: accept URLs with "query strings", that get included in local filename
- (experimental) if missing, look for resources in winetricks cache before attempting download (#1963)
- Add POL_Wine_DelOverrideDLL, POL_Wine_DelAppOverrideDLL (#1989)
- POL_Wine_OverrideDLL and POL_Wine_AppOverrideDLL cleanups
- Support creation of desktop menu entries (#1943). POL_Shortcut needs
  an extra parameter, categories, for the entry to be created. And
  similarly to NO_DESKTOP_ICON setting, a NO_MENU_ICON can be set to
  TRUE to prevent desktop menu entries from being created.

-> 4.1.9
- POL_System_is32bit
- Improved package menu
- Added : POL_Download_GetSize, POL_System_HomeSpaceLeft
- Fixed bug : #1446
- POL_Wine_PrefixCreate(): wait for wineserver to finish instead of hardcoded 3s wait (#1486)
- Fix bug in POL_Download_GetSize when redirected
- Fix options parsing bug in POL_System_CheckFS
- Log when NO_FSCHECK TRUE has been used
- Tried to make it clearer that no-cd patches may be needed to get legal programs to work
- Handle prefix overwrite in POL_Wine_PrefixCreate (Overwrite/Erase/Abort)
- POL_Download_Resource(): fix for $3 and $4 mixed up (#1651)
- VersionLower: fix cases when 3rd component in version strings is missing ("1.4") or dot in development string ("1.4-dos_support_0.6")
- Fixes to lib.playonlinux.VersionLower()
- Set_Desktop: support Wine 1.5.16+ (#1710)
- Try to make the purpose of install filters more obvious (#1652)

-> 4.1.8
- Wine issue fixed on OSX 10.8.2 and 10.7.5
- Make "missing package" messages a bit more explicit
- Fix typo in POL_Wine_OverrideDLL
- Don't log regedit parameter if not a file
- irc client: changed reg.sub() to not depend on flag parameter (Python 2.7+)

-> 4.1.7
- Search bar
- Bug fix
- DONT_KILL_ON_EXIT setting
- additional debug messages in POL_Download and POL_Download_Resource
- additional debug messages in POL_Call
- Many stability and UI improvements

-> 4.1.6
- Critical fixes

-> 4.1.5
- POL_Debug_* cookie forgotten 
- Bug in update process

-> 4.1.4
- Timer: read modification date instead of listing directories (CPU usage)
- CheckGL command
- Bug 1076, quote disappear after configure menu fixed
- Clickable links on install window
- Better (and clearer) debug windows
- Small fix for OSX 10.8
- NEW: PlayOnLinux --run supports extensions! 
- POL_Shortcut_Document : Researches
- New POL_SetupWindow_browse GUI
- POL_SetupWindow_browse supports filters
- Ask to redownload when failed
- Minor: bug 1053
- Hypertext link in IRC
- Some improvement in IRC
- PlayOnLinux user agent
- A new method to manage setup window (cleaner code)
- Added POL_SimpleAlert
- Exiting POL will kill all opened bash scripts
- Updating the GIT is possible from the GUI
- Several UIs improvements
- Added POL_Die(), POL_Restart(), POL_System_RegisterPID()
- Harden POL_Download, POL_Download_Resource, POL_System_wget, POL_System_CopyDirectory against errors (disk full?)
- Create hardened commands POL_System_unzip, POL_System_7z, POL_System_tar, POL_System_cabextract, POL_System_unrar

-> 4.1.3
- Real time refreshing
- New install window
- 64x64 icons support
- 24bits icon support
- regedit alone does not log anything

python/lib/playonlinux.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright (C) 2007-2010 PlayOnLinux Team

import Variables, os, string
import subprocess, shlex, pipes, wx

def get_foreground_colour():
    return wx.SystemSettings.GetColour( wx.SYS_COLOUR_BTNTEXT )

def get_foreground_hover_colour():
    return wx.SystemSettings.GetColour( wx.SYS_COLOUR_3DHILIGHT )

def get_background_colour():
    return wx.SystemSettings.GetColour( wx.SYS_COLOUR_BACKGROUND )

def get_background_test_colour():
    return wx.Colour(90,90,222)

def winpath(script, path):
    #path=os.path.realpath(path)
    if(path[0] != "/"):
        path=os.environ["WorkingDirectory"]+"/"+path
    path = os.path.realpath(path)
    pref = getPrefix(script)
    ver = GetSettings('VERSION',pref)
    arch = GetSettings('ARCH',pref)
    if(arch == ""):
        arch="x86"
    if(ver == ""):
        return(os.popen("env WINEPREFIX='"+os.environ["POL_USER_ROOT"]+"/wineprefix/"+pref+"/' 'wine' winepath -w '"+path+"'").read().replace("\n","").replace("\r",""))
    else:
        return(os.popen("env WINEPREFIX='"+os.environ["POL_USER_ROOT"]+"/wineprefix/"+pref+"/' '"+os.environ["POL_USER_ROOT"]+"/wine/"+Variables.os_name+"-"+arch+"/"+ver+"/bin/wine' winepath -w '"+path+"'").read().replace("\n","").replace("\r",""))

def open_document(path, ext):
    script = GetSettings(ext, '_EXT_')
    if(script == ""):
        wx.MessageBox(_("There is nothing installed to run .{0} files.").format(ext),os.environ["APPLICATION_TITLE"], wx.OK)
    else:
        try:
            subprocess.Popen(["bash", Variables.playonlinux_env+"/bash/run_app", script.encode("utf-8","replace"), winpath(script.encode("utf-8","replace"), path.encode("utf-8","replace"))])
        except:
            subprocess.Popen(["bash", Variables.playonlinux_env+"/bash/run_app", script, winpath(script, path)])

def GetWineVersion(game):
    cfile = Variables.playonlinux_rep+"shortcuts/"+game
    fichier = open(cfile,"r").readlines()
    i = 0
    line = ""
    while(i < len(fichier)):
        fichier[i] = fichier[i].replace("\n","")
        if("PATH=" in fichier[i] and "WineVersions" in fichier[i]):
            line = fichier[i].replace("//","/")
        i += 1

    if(line == ""):
        version = "System"
    else:
        version=line.replace("PATH=","").replace("\"","").replace(Variables.playonlinux_rep,"").replace("//","/")
        version = string.split(version,"/")
        version = version[1]

    return(version)

def GetSettings(setting, prefix='_POL_'):
    if(prefix == "_POL_"):
        cfile = Variables.playonlinux_rep+"/playonlinux.cfg"
    elif(prefix == "_EXT_"):
        cfile = Variables.playonlinux_rep+"/extensions.cfg"
    elif(prefix == "_SHV_"):
	cfile = Variables.playonlinux_rep+"/shelves.cfg"
    else:
        cfile = Variables.playonlinux_rep+"/wineprefix/"+prefix+"/playonlinux.cfg"

    try:
        fichier = open(cfile,"r").readlines()
    except:
        return("")

    i = 0
    line = ""
    while(i < len(fichier)):
        fichier[i] = fichier[i].replace("\n","")
        if(setting+"=" in fichier[i]):
            line = fichier[i]
            break
        i += 1
    try:
        line = string.split(line,"=")
        del line[0]
        return(string.join(line,"="))
    except:
        return("")

def SetSettings(setting, value, prefix='_POL_'):
    if(prefix == "_POL_"):
        cfile = Variables.playonlinux_rep+"/playonlinux.cfg"
    elif(prefix == "_EXT_"):
        cfile = Variables.playonlinux_rep+"/extensions.cfg"
    elif(prefix == "_SHV_"):
	cfile = Variables.playonlinux_rep+"/shelves.cfg"
    else:
        cfile = Variables.playonlinux_rep+"/wineprefix/"+prefix+"/playonlinux.cfg"

    try:
        fichier = open(cfile,"r").readlines()
    except:
        pass
    else:
	if(prefix == "_SHV_"):
	    fichier = open(cfile,"w")
	    fichier.write(value)
	    fichier.close()
	else:
	    i = 0
	    line = []
	    found = False
	    while(i < len(fichier)):
		fichier[i] = fichier[i].replace("\n","")
		if(setting+"=" in fichier[i]):
		    line.append(setting+"="+value)
		    found = True
		else:
		    line.append(fichier[i])
		i += 1
	    if(found == False):
		line.append(setting+"="+value)

	    try:
		fichier_write = open(cfile,"w")
	    except IOError:
		pass
	    else:
		i = 0
		while(i < len(line)): # On ecrit
		    fichier_write.write(line[i]+"\n")
		    i+=1

def DeleteSettings(setting, prefix='_POL_'):
    if(prefix == "_POL_"):
        cfile = Variables.playonlinux_rep+"/playonlinux.cfg"
    elif(prefix == "_EXT_"):
        cfile = Variables.playonlinux_rep+"/extensions.cfg"
    elif(prefix == "_SHV_"):
	cfile = Variables.playonlinux_rep+"/shelves.cfg"
    else:
        cfile = Variables.playonlinux_rep+"/wineprefix/"+prefix+"/playonlinux.cfg"

    fichier = open(cfile,"r").readlines()
    i = 0
    line = []
    found = False
    while(i < len(fichier)):
	fichier[i] = fichier[i].replace("\n","")
	if(setting+"=" not in fichier[i]):
	    line.append(fichier[i])
	i += 1

    fichier_write = open(cfile,"w")

    i = 0
    while(i < len(line)): # On ecrit
	fichier_write.write(line[i]+"\n")
	i+=1


def getLog(game):
    cfile = Variables.playonlinux_rep+"shortcuts/"+game
    try:
        fichier = open(cfile,"r").readlines()
    except:
        return None

    for line in fichier:
        line = line.replace("\n","")
        if('#POL_Log=' in line):
            line = string.split(line,"=")
            return(line[1])
    return None

def GetDebugState(game):
    cfile = Variables.playonlinux_rep+"shortcuts/"+game
    try:
        fichier = open(cfile,"r").readlines()
    except:
        return True

    for line in fichier:
        line = line.replace("\n","")
        if('export WINEDEBUG=' in line):
            if(line == 'export WINEDEBUG="-all"'):
                return False
            else:
                return True
    return False

def SetDebugState(game, prefix, state):
    cfile = Variables.playonlinux_rep+"shortcuts/"+game
    try:
        fichier = open(cfile,"r").readlines()
    except:
        return False

    lines = []
    for line in fichier:
        line = line.replace("\n","")
        if('export WINEDEBUG=' in line):
            if(state == True):
                line = 'export WINEDEBUG="%s"' % GetSettings('WINEDEBUG', prefix)
            else:
                line = 'export WINEDEBUG="-all"'
        lines.append(line)

    fichier_write = open(cfile,"w")

    i = 0
    while(i < len(lines)): # On ecrit
        fichier_write.write(lines[i]+"\n")
        i+=1

def keynat(string):
    r'''A natural sort helper function for sort() and sorted()
    without using regular expressions or exceptions.

    >>> items = ('Z', 'a', '10th', '1st', '9')
    >>> sorted(items)
    ['10th', '1st', '9', 'Z', 'a']
    >>> sorted(items, key=keynat)
    ['1st', '9', '10th', 'a', 'Z']

    Borrowed from http://code.activestate.com/recipes/285264/#c6
    by paul clinch.

    License is the PSF Python License, http://www.python.org/psf/license/ (GPL compatible)
    '''
    it = type(1)
    r = []
    for c in string:
        if c.isdigit():
            d = int(c)
            if r and type( r[-1] ) == it:
                r[-1] = r[-1] * 10 + d
            else:
                r.append(d)
        else:
            r.append(c.lower())
    return r

def open_folder(software, cmd="xdg-open"):
    read = open(Variables.playonlinux_rep+"shortcuts/"+software,"r").readlines()

    if not len(read):
        return

    i = 0;
    while(i < len(read)):
        if("cd \"" in read[i]):
            break
        i += 1

    if len(read) == (i):
        return

    AppDir = read[i][4:-2].replace("//","/")
    if AppDir != "":
        if(os.environ["POL_OS"] == "Mac"):
            subprocess.call(["open", AppDir])
        else:
            subprocess.call([cmd, AppDir])

def open_folder_prefix(software):
    AppDir = os.environ["POL_USER_ROOT"]+"/wineprefix/"+software
    if AppDir != "":
        if(os.environ["POL_OS"] == "Mac"):
            subprocess.call(["open", AppDir])
        else:
            subprocess.call(["xdg-open", AppDir])

def VersionLower(version1, version2):
    version1 = string.split(version1, "-")
    version2 = string.split(version2, "-")

    try:
        if(version1[1] != ""):
            dev1 = True
    except:
        dev1 = False

    try:
        if(version2[1] != ""):
            dev2 = True
    except:
        dev2 = False

    if(version1[0] == version2[0]):
        if(dev1 == True and dev2 == False):
            return True
        else:
            return False

    version1 = [ int(digit) for digit in string.split(version1[0],".") ]
    while len(version1) < 3:
        version1.append(0)

    version2 = [ int(digit) for digit in string.split(version2[0],".") ]
    while len(version2) < 3:
        version2.append(0)

    if(version1[0] < version2[0]):
        return True
    elif(version1[0] == version2[0]):
        if(version1[1] < version2[1]):
            return True
        elif(version1[1] == version2[1]):
            if(version1[2] < version2[2]):
                return True
            else:
                return False
        else:
            return False
    else:
        return False

def convertVersionToInt(version): # Code par MulX en Bash, adapte en python par Tinou
    #rajouter pour les vesions de dev -> la version stable peut sortir
    #les personnes qui utilise la version de dev sont quand même informé d'une MAJ
    #ex 3.8.1 < 3.8.2-dev < 3.8.2
    print "Deprecated !"
    if("dev" in version or "beta" in version or "alpha" in version or "rc" in version):
        version = string.split(version,"-")
        version = version[0]
        versionDev = -5
    else:
        versionDev = 0

    version_s = string.split(version,".")
    #on fait des maths partie1 elever au cube et multiplier par 1000
    try:
        versionP1 = int(version_s[0])*int(version_s[0])*int(version_s[0])*1000
    except:
        versionP1 = 0
    try:
        versionP2 = int(version_s[1])*int(version_s[1])*100
    except:
        versionP2 = 0
    try:
        versionP3 = int(version_s[2])*10
    except:
        versionP3 = 0
    return(versionDev + versionP1 + versionP2 + versionP3)

def getPrefix(shortcut): # Get prefix name from shortcut
    if(os.path.isdir(os.environ["POL_USER_ROOT"]+"/shortcuts/"+shortcut)):
        return ""

    fichier = open(os.environ["POL_USER_ROOT"]+"/shortcuts/"+shortcut,'r').read()
    fichier = string.split(fichier,"\n")
    i = 0
    while(i < len(fichier)):
        if("export WINEPREFIX=" in fichier[i]):
            break
        i += 1

    try:
        prefix = string.split(fichier[i],"\"")
        prefix = prefix[1].replace("//","/")
        prefix = string.split(prefix,"/")

        if(os.environ["POL_OS"] == "Mac"):
            index_of_dotPOL = prefix.index("PlayOnMac")
            prefix = prefix[index_of_dotPOL + 2]
        else:
            index_of_dotPOL = prefix.index(".PlayOnLinux")
            prefix = prefix[index_of_dotPOL + 2]
    except:
        prefix = ""

    return prefix


def getArgs(shortcut): # Get prefix name from shortcut
    if(os.path.isdir(os.environ["POL_USER_ROOT"]+"/shortcuts/"+shortcut)):
        return ""

    fichier = open(os.environ["POL_USER_ROOT"]+"/shortcuts/"+shortcut,'r').read()
    fichier = string.split(fichier,"\n")
    i = 0
    while(i < len(fichier)):
        if("POL_Wine " in fichier[i]):
            break
        i += 1

    try:
        args = shlex.split(fichier[i])[2:-1]
        #print args
        args = " ".join([ pipes.quote(x) for x in args])
        #print args
    except:
        args = ""

    return args

def Get_versions(arch='x86'):
    installed_versions = os.listdir(Variables.playonlinux_rep+"/wine/"+Variables.os_name+"-"+arch+"/")
    installed_versions.sort(key=keynat)
    installed_versions.reverse()
    try:
        installed_versions.remove("installed")
    except:
        pass
    return installed_versions

def Get_Drives():
    pref = os.listdir(Variables.playonlinux_rep+"/wineprefix/")
    pref.sort()
    return pref

def Get_Shelves():
    return open(os.environ["POL_USER_ROOT"]+"/shelves.cfg").readlines()

def SetWinePrefix(game, prefix):
    cfile = Variables.playonlinux_rep+"shortcuts/"+game
    fichier = open(cfile,"r").readlines()
    i = 0
    line = []
    while(i < len(fichier)): # On retire l'eventuel
        fichier[i] = fichier[i].replace("\n","")
        if("export WINEPREFIX=" not in fichier[i] or "/wineprefix/" not in fichier[i]):
            line.append(fichier[i])
        else:
            line.append("export WINEPREFIX=\""+Variables.playonlinux_rep+"/wineprefix/"+prefix+"\"")
        i += 1

    fichier_write = open(cfile,"w")

    i = 0
    while(i < len(line)): # On ecrit
        fichier_write.write(line[i]+"\n")
        i+=1


def writeArgs(game, args):
    cfile = Variables.playonlinux_rep+"shortcuts/"+game
    fichier = open(cfile,"r").readlines()
    i = 0
    line = []

    while(i < len(fichier)): # On retire l'eventuel
        fichier[i] = fichier[i].replace("\n","")
        if("POL_Wine " not in fichier[i]):
            line.append(fichier[i])
        else:
            try:
                old_string = shlex.split(fichier[i])
                new_string = shlex.split(str(args))
                new_string = old_string[0:2] + new_string
                new_string = " ".join([ pipes.quote(x) for x in new_string])

                new_string = new_string+' "$@"'
                line.append(new_string)
            except:
                line.append(fichier[i])
        i += 1

    fichier_write = open(cfile,"w")
    i = 0
    while(i < len(line)): # On ecrit
        fichier_write.write(line[i]+"\n")
        i+=1

def POL_Open(arg):
    if(os.environ["POL_OS"] == "Mac"):
        subprocess.Popen(["open", arg])
    else:
        subprocess.Popen(["xdg-open", arg])

def POL_Error(message):
    wx.MessageBox(message,_("{0} error").format(os.environ["APPLICATION_TITLE"]))

def SetMenuBitmap(menuitem, path):
    if(os.path.exists(path)):
	bitmap = wx.Bitmap(path)
	menuitem.SetBitmap(bitmap)

def AddMenuItem(parent, menuId, label, pathtobitmap):
    newMenuItem = wx.MenuItem(parent, menuId, label)
    if(os.path.exists(pathtobitmap)):
	SetMenuBitmap(newMenuItem, pathtobitmap)
    parent.AppendItem(newMenuItem)
    return newMenuItem
